### PR TITLE
Add Wiki retreval service

### PR DESCRIPTION
**Pull request by PengfeiHePower**
---
name: Pull Request
about: Create a pull request
---
## Description

wiki.py implements Wikipedia retrieval including text, category list, infobox, image and table.
wiki_test.py implement unit tests.

## Checklist

Please check the following items before code is ready to be reviewed.

- [ ]  Code has passed all tests
- [ ]  Docstrings have been added/updated in Google Style
- [ ]  Documentation has been updated
- [ ]  Code is ready for review

### CODE REVIEW

File: src/agentscope/service/web/wiki.py [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+    search_params = {
+        "action": "query",
+        "list": "search",
+        "srsearch": entity,
+        "format": "json",
+    }
+
+    search_data = wiki_api(search_params)
+
+    if "query" in search_data and search_data["query"]["search"]:
```
Comment (ZiTao-Li) [inline src/agentscope/service/web/wiki.py:35]: is it possible that "search" does not exist in the search_data["query"]?

**Review by ZiTao-Li — COMMENTED** [review]
See the inline comment. Other LGTM

File: src/agentscope/service/web/wiki.py (hunk 1 above)
Comment (PengfeiHePower) [reply src/agentscope/service/web/wiki.py:35]: The structure of search_data is: {'query':{'searchinfo':xxx, 'search':xxx,...}} so there is always a 'search' in 'query'.

**Review by DavdGao — COMMENTED** [review]
I'm confused about the functions provided in this PR as follows:

1. Why do we provide `wiki_get_infobox`, `wiki_get_all_wikipedia_tables`, `wiki_get_page_images_with_captions` and `wiki_get_page_content_by_paragraph` respectively rather than combining them into one function that provides wikipedia search functionality, and returns all available information? For example: 
```python
{
    "infobox": ...
    "content": ...  # the tables should be in the content
}
```

If I was a user and want to search in wikipedia, a simple `wikipedia_search` function will be enough, and too much categorization can create a barrier to understanding. For example: 
- What do the `infobox` and `tables` stand for?
- For users, how to decide which function to use?

2. Partial functionality is repetitive with the service function `digest_webpage`. Try to reuse it rather than re-implementation

3. I'm curious about how do we handle the fuzzy search in wikipedia. For example, sometimes wikipedia will return a candidate list rather than skip to the corresponding web page as follows.

image

File: src/agentscope/service/web/wiki.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
+# -*- coding: utf-8 -*-
+"""
+Search contents from WikiPedia,
+including texts, categories, infotable, table,...
+"""
+
+import re
+
+# import requests
```
[the remark this answers is not in the document]
Comment (PengfeiHePower) [reply src/agentscope/service/web/wiki.py:9]: Fixed

File: src/agentscope/service/web/wiki.py [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+from bs4 import BeautifulSoup
+
+from agentscope.service.service_response import (
+    ServiceResponse,
+    ServiceExecStatus,
+)
+from agentscope.utils.common import requests_get
+
+
+def wiki_api(params: dict) -> dict:
```
[the remark this answers is not in the document]
Comment (PengfeiHePower) [reply src/agentscope/service/web/wiki.py:19]: Fixed

File: src/agentscope/service/web/wiki.py [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+    entity: str,
+    max_members: int = 1000,
+    limit_per_request: int = 500,
+) -> ServiceResponse:
+    """Function to retrieve category members from Wikipedia:Category pages
+
+    Args:
+        entity (str): searching keywords
+        max_members (int): maximum number of members to output
+        limit_per_request (int): number of members retrieved per quest
```
[the remark this answers is not in the document]
Comment (PengfeiHePower) [reply src/agentscope/service/web/wiki.py:71]: Fixed

**Comment by PengfeiHePower** [comment]
> I'm confused about the functions provided in this PR as follows:
> 
> 1. Why do we provide `wiki_get_infobox`, `wiki_get_all_wikipedia_tables`, `wiki_get_page_images_with_captions` and `wiki_get_page_content_by_paragraph` respectively rather than combining them into one function that provides wikipedia search functionality, and returns all available information? For example:
> 
> ```python
> {
>     "infobox": ...
>     "content": ...  # the tables should be in the content
>     "image": ...
> }
> ```
> 
> If I was a user and want to search in wikipedia, a simple `wikipedia_search` function will be enough, and too much categorization can create a barrier to understanding. For example:
> 
> * What do the `infobox` and `tables` stand for?
> * For users, how to decide which function to use?
> 
> 2. The functionality partly overlaps with that of the service function digest_webpage. Try to reuse it rather than reimplementing.
> 3. I wonder how do we handle the fuzzy search in wikipedia. For example, sometimes wikipedia will return a candidate list rather than skip to the corresponding web page as follows.
> 
> image

1. I have added a new function "wiki_page_retrieval" which includes all types of information (output as a dictionary) excluding the category list. This is because the category list is a totally different type of content.
2. We would like to maintain a key-val structure for contents in 'infobox', but the 'parse_html_to_text' function in web_digest only provide plain texts and it it not convenient to reuse.
3. We have provided a function '_check_entity_existence' which will first check if the entity exists in Wiki. If the entity does not have a page (as in your example), it will return the first 5 entities listed below (i.e. Tong Dawei, Gao Lu,... in your example).

**Review by DavdGao — APPROVED** [review]
lgtm

**Pull request merged**